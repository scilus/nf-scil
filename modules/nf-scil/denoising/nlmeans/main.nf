
process DENOISING_NLMEANS {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_1.5.0.sif':
        'scilus/scilus:1.5.0' }"

    input:
    tuple val(meta), path(image), path(mask)

    output:
    tuple val(meta), path("*_denoised.nii.gz"), emit: image
    path "versions.yml"                       , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def args = []
    if (mask) args += ["--mask $mask"]

    """
    scil_run_nlmeans.py $image ${prefix}_denoised.nii.gz 1 ${args.join(" ")}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(scil_get_version.py 2>&1)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    
    """
    scil_run_nlmeans.py -h

    touch ${prefix}_denoised.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        : \$(scil_get_version.py 2>&1)
    END_VERSIONS
    """
}
