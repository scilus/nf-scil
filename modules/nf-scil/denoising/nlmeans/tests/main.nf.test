nextflow_process {

    name "Test Process DENOISING_NLMEANS"
    script "../main.nf"
    process "DENOISING_NLMEANS"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfcore"
    tag "denoising"
    tag "denoising/nlmeans"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-scil/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "raw_b0.zip", "raw_segmentation.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("denoising - nlmeans") {

        when {
            process {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        b0: it.simpleName == "raw_b0"
                        segmentation: it.simpleName == "raw_segmentation"
                    }
                ch_b0 = ch_split_test_data.b0.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/b0.nii.gz")
                    ]
                }
                ch_mask = ch_split_test_data.segmentation.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/brainmask/slices/axial.nii.gz")
                    ]
                }
                input[0] = ch_b0
                    .join(ch_mask)
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

}
