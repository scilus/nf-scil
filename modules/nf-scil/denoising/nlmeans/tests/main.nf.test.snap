{
    "denoising - nlmeans": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,a22cff5c79101e8d2e19cf83c7e1d3df"
                    ]
                ],
                "1": [
                    "versions.yml:md5,e5c9936e9d01b8503f706aecd0a39027"
                ],
                "image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,a22cff5c79101e8d2e19cf83c7e1d3df"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e5c9936e9d01b8503f706aecd0a39027"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-27T02:12:39.397869"
    },
    "denoising - nlmeans - withmask": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b5c8b52c64427a2b7a6645ba085fa61a"
                    ]
                ],
                "1": [
                    "versions.yml:md5,e5c9936e9d01b8503f706aecd0a39027"
                ],
                "image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_denoised.nii.gz:md5,b5c8b52c64427a2b7a6645ba085fa61a"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,e5c9936e9d01b8503f706aecd0a39027"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0-rc1",
            "nextflow": "24.04.2"
        },
        "timestamp": "2024-07-27T02:12:51.206297"
    }
}